README
********************************************************************

As of 6/27/213 Scalpel has been released under the Apache 2.0 License
and the source is available at The Sleuth Kit github repository.

Bug reports, comments, complaints, and feature requests should be
directed to sleuthkit-users@lists.sourceforge.net.

********************************************************************

Scalpel is a file carving and indexing application that runs on Linux
and Windows.  The first version of Scalpel, released in 2005, was
based on Foremost 0.69. There have been a number of internal releases
since the last public release, 1.60, primarily to support our own
research.  The newest public release v2.0, has a number of additional
features, including:

o minimum carve sizes.

o multithreading for quicker execution on multicore CPUs.

o asynchronous I/O that allows disk operations to overlap with pattern
matching--this results in a substantial performance improvement.

o regular expression support for headers/footers.

o embedded header/footer matching for better processing of structured
file types that may contain embedded files.

o for advanced users, support for massively-threaded execution on
Graphics Processing Units (GPUs).  This feature is available only on
Linux and requires installation of the NVIDIA CUDA SDK, modification
of scalpel.h to enable the GPU threading mode, and compilation with
the CUDA toolchain.  Our implementation also requires an NVIDIA GPU
with compute capability >= 1.2, so older CUDA-capable cards probably
won't work.  The NVIDIA GTX 260 is relatively inexpensive and powerful
and has the appropriate compute capability.  The GPU-enhanced version
of Scalpel is able to do preview carving at rates that exceed the disk
bandwidth of most file servers, so for big jobs, it may be worth the
extra effort required to use this feature.  Note that regular
expression-based headers and footers are NOT currently supported when
GPU acceleration is in use!  We might address this in a future
release.

Scalpel performs file carving operations based on patterns that
describe particular file or data fragment "types".  These patterns may
be based on either fixed binary strings or regular expressions.  A
number of default patterns are included in the configuration file
included in the distribution, "scalpel.conf".  The comments in the
configuration file explain the format of the file carving patterns
supported by Scalpel.

Important note: The default configuration file, "scalpel.conf", has
all supported file patterns commented out--you must edit this file
before running Scalpel to activate some patterns.  Resist the urge to
simply uncomment all file carving patterns; this wastes time and will
generate a huge number of false positives.  Instead, uncomment only
the patterns for the file types you need.

Scalpel options are described in the Scalpel man page, "scalpel.1".
You may also execute Scalpel w/o any command line arguments to see a
list of options.

NOTE: Compilation is necessary on Unix platforms and on Mac OS X.  For
Windows platforms, a precompiled scalpel.exe is provided.  If you do
wish to recompile Scalpel on Windows, you'll need a mingw (gcc)
setup. Scalpel will not compile using Visual Studio C compilers.  Note
that our compilation environment for Windows is currently 32-bit; we
haven't tested on the 64-bit version of mingw, but will address this
int the future.

COMPILE INSTRUCTIONS ON SUPPORTED PLATFORMS:

Linux/Mac OS X:    ./configure and then make

Windows:           cd to src directory and then:

	           mingw32-make -f Makefile.win

and enjoy.  If you want to install the binary and man page in a more
permanent place, just copy "scalpel" (or "scalpel.exe") and
"scalpel.1" to appropriate locations, e.g., on Linux, "/usr/local/bin"
and "/usr/local/man/man1", respectively.  On Windows, you'll also need
to copy the pthreads and tre regular expression library dlls into the
same directory as "scalpel.exe".


OTHER SUPPORTED PLATFORMS

We are not currently supporting Scalpel on Unix variants other than
Linux. Go ahead and try a ./configure and make and see what happens,
but be sure to do thorough testing before using Scalpel in production
work.  If you are interested in supporting a version of Scalpel on an
alternate platforms, please contact us.  If you are interested in
supporting a GPU-enhanced version of Scalpel on Windows, we are also
interesting in hearing from you.


LIMITATIONS:

Carving Windows physical and logical device files (e.g.,
\\.\physicaldrive0 or \\.\c:) isn't currently supported because it
requires us to rewrite some portions of Scalpel to use Windows file
I/O functions rather than standard Unix calls.  This may be supported
in a future release.

Block map features are currently disabled, as we are rewriting this
subsystem to enhance interoperability with the Sleuthkit.  An improved
version of the block map features will return in a subsequent release.
The -s command line option ("skip") has been removed and will be
replaced with a more robust facility in the next major release.


DEPENDENCIES:

Scalpel uses the POSIX threads library.  On Win32, Scalpel is
distributed with the Pthreads-win32 - POSIX Threads Library for Win32,
which is Copyright(C) 1998 John E. Bossom and Copyright(C) 1999,2005
by Pthreads-win32 contributors.  This library is licensed under the LGPL.

Scalpel for Win32 uses the tre regular expression library and is
distributed with tre-0.7.5, which is licensed under the LGPL.

Cheers,

--Golden and Vico.









Changelog
1.4:	First public beta release.
1.5:	Added "-b" option for complete carving compability with
	foremost 0.69.
1.51:	Fixed a problem with Scalpel not carving all files near the end of
	some large images. The locations of the files was detected, but an
	optimization to minimize reads on pass 2 caused the files not to
	be written (and not to be noted in the Scalpel log).  Also, more
	work to bring the Win32 version into compliance.  Fixed problem
	with Win32 version reporting "Couldn't measure size of image file"
	for large images--required ftello64()/fseeko64() under mingw.
	40+GB images have now been carved successfully under Win32.  Fixed
	format of offset in log file for large offsets under Win32; the
	printf() implementation in mingw under Windows doesn't support
	"%llu" for unsigned long long integers--needs "%I64u".  Blech.
1.52	Foremost 0.69 didn't discover overlapping headers/footers.  If you want
	this behavior, use "-r" on the command line.  Scalpel's default
	behavior is now to discover all headers/footers, even if they
	overlap.  Different Quicktime needles (from the 'file' command's
	magic file) are now used in the configuration file, because the
	default Foremost needles didn't seem to match valid Quicktime
	files.  The needles are still weak, in that they cause a bunch of
	false positives.  The configuration file now uses more
	reasonable (in my opinion) maximum carve sizes.  By default, all
	file types are still commented out.
1.53	Verbose mode implemented.  "-v" outputs copious (too much?) information
	about what Scalpel is doing as files are carved.  Mac OS X is now
	supported, including the use of raw block devices
	(e.g., /dev/disk0) as carve targets.  Use "make bsd" for an OS X
	compilation.  NOTE: As of version 2.0, this support is now
	deprecated, but may be reintroduced in the future.
1.54	Maximum size for carved files is now 18 exabytes. Minor changes to
	Makefile. Fixed minor formatting problem in audit log.  Scalpel
	man page created.  Eliminated duplicate error messages when
	targets could not be opened.  Thanks to John Vigo for additional
	MAC OS X testing on this release.
1.60	Changes in this release include:

	o Some of the headers and footers in the default
	'scalpel.conf' file have been improved.  In particular, the
	GIF89 footer was made less restrictive; GIF89 files were
	encountered in the wild for which the legacy Foremost footer
	was too restrictive.

	o REVERSE carving semantics now fixed as "helpers.c"
	functions continue to evolve from their Foremost roots; this
	is mostly useful for PDF carving and when some carved files
	appear to be truncated--e.g., many times GIF files will
	contain the footer string internally and with FORWARD carving,
	will be truncated.  REVERSE carving can help, but at the
	expense of occasionally carving overly large files.

	o Removed obsolete option "-f".

	o Maximum number of files that can be carved per file type is
	now 18,446,744,073,709,551,616.  This is also the maximum
	number of files of all combined types that can be carved in
	a single run.  This is unlikely to be a problematic upper
	bound in the near term.

	o The "-q" command line option is now available to force
	block-aligned carving; only files with headers that begin on
	block boundaries are carved.  The block size for this mode is
	user-specified.

	o The "-r" option was not accepted in recent releases due to
	an omission in Scalpel's command line argument parsing.  This
	option isn't generally used except for compatibility testing
	with Foremost 0.69 (it actually forces Scalpel to mimic some
	buggy Foremost behavior) and should be considered deprecated.
	In a future release of Scalpel, it will be removed.

	o Scalpel now organizes carved files into subdirectories by
	type (and for each type, additional subdirectories are created
	with a maximum of MAX_FILES_PER_SUBDIRECTORY [default: 5000]
	files each).  This will allow you to more comfortably explore
	the set of carved files using the Windows or Linux file
	explorers without having them choke on large subdirectories.
	The files are organized within the Scalpel output directory
	into subdirectories with the following naming convention:

	suffix-rule#-counter
	suffix-rule#-counter
	...
	suffix-rule#-counter

	where 'suffix' matches the suffix specified in the Scalpel
	configuration file (e.g, "jpg"), 'rule#' is the index of the
	corresponding rule in the configuration file (with the first
	uncommented rule being 0), and 'counter' being incremented after
	every MAX_FILES_PER_SUBDIRECTORY files corresponding to a rule are
	carved.

	You can turn off this default organization by using the "-O"
	command line flag (then Scalpel acts as in previous releases
	and dumps all carved files into a single output directory).  

	o The "-p" performs a "preview" carve, creating an audit file
	indicating which files would be carved, but without performing
	carving operations.  This substantially increases execution speed
	and also supports "in-line" carving, when used in conjunction with
	our custom FUSE filesystem.

	o A memory corruption error when Scalpel had large numbers of
	files open (typically, when a large number of file types were
	specified in the configuration file) was corrected.  Thanks to
	both Doug Koster and Peter Theobald for pointing out this error.

	o In this release some unused functions were removed
	from the Scalpel source.

	o A bug which arose for case-insensitive header/footer
	specifications in the legacy Foremost Boyer-Moore string
	search initialization was corrected.  Thanks to Doug Koster
	for pointing out this bug.

	o Fixed (slightly) garbled invocation string in audit file; this
	resulted from an unitialized string in legacy Foremost code.

	o The "-d", "-m", "-t", and "-u" options are experimental and are
	part of evolving support for interaction between Scalpel and
	external tools, with the goal being better support for recovery
	of fragmented files.   Please contact the author -->

src/scalpel.h
/*
Copyright (C) 2013, Basis Technology Corp.
Copyright (C) 2007-2011, Golden G. Richard III and Vico Marziale.
Copyright (C) 2005-2007, Golden G. Richard III.
*
Written by Golden G. Richard III and Vico Marziale.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
*
http://www.apache.org/licenses/LICENSE-2.0
*
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
Thanks to Kris Kendall, Jesse Kornblum, et al for their work
on Foremost. Foremost 0.69 was used as the starting point for
Scalpel, in 2005.
*/

#ifndef SCALPEL_H
#define SCALPEL_H
#define SCALPEL_VERSION    "2.1"

//#define GPU_THREADING
#ifndef SINGLE_THREADING
#define MULTICORE_THREADING
#endif
#define USE_FAST_STRING_SEARCH

#define _USE_LARGEFILE              1
#define _USE_FILEOFFSET64           1
#define _USE_LARGEFILE64            1
#define _LARGEFILE_SOURCE           1
#define _LARGEFILE64_SOURCE         1
#define _FILE_OFFSET_BITS           64

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <ctype.h>
#include <sys/stat.h>
#include <time.h>
#include <errno.h>
#include <signal.h>
#include <limits.h>
#include <dirent.h>
#include <stdarg.h>
#include <math.h>
#include <pthread.h>
#include <semaphore.h>
#include <sys/timeb.h>
#include <sys/time.h>

//C++ STL headers
#include <exception>
#include <stdexcept>
#include <string>
#include <sstream>

//scalpel headers
#include "base_name.h"
#include "prioque.h"
#include "syncqueue.h"
#include "common.h"
#include "types.h"

#include "input_reader.h"


#ifdef __APPLE__
#define __UNIX
#include <sys/ttycom.h>
#include <sys/param.h>
#include <sys/ioctl.h>
#include <libgen.h>
#include <tre/regex.h>
// off_t on Mac OS X is 64 bits
#define off64_t  off_t
#endif /* ifdef __APPLE__ */

#ifdef __linux
#define __UNIX
#include <linux/hdreg.h>
#include <libgen.h>
#include <error.h>
#include <tre/regex.h>
#endif /* ifdef __linux */

#if defined ( _WIN32)
#include <windows.h>
#include <sys/timeb.h>
#include "regex.h"
#define gettimeofday(A, B) QueryPerformanceCounter(A)
#define sleep    Sleep
#define  snprintf         _snprintf
#define lstat(A,B)      stat(A,B)
#include <libgen.h>
//char *basename (char *path);
extern char *optarg;
extern int optind;
int getopt (int argc, char *const argv[], const char *optstring);

#ifdef __MINGW32__
#define realpath(A,B)    _fullpath(B,A,PATH_MAX)
#endif

#ifdef __CYGWIN32__
#define realpath(A,B) \
  (getenv ("CYGWIN_USE_WIN32_PATHS") \
   ? (cygwin_conv_to_full_win32_path ((A), (B)), B) \
   : realpath ((A), (B)))
#endif
#else // ! defined(WIN32)
#include <sys/mount.h>
#define gettimeofday_t struct timeval
#endif // ! defined(WIN32)

#define SEARCHTYPE_FORWARD      0
#define SEARCHTYPE_REVERSE      1
#define SEARCHTYPE_FORWARD_NEXT 2

// LARGEST_REGEXP_OVERLAP specifies the largest regular expression overlap
// across the boundaries of SIZE_OF_BUFFER-sized chunks of the disk image.
//  This is also used internally as the maximum "size" of a regular expression
// and affects the mininum disk image size that can be processed.  Large
// values will have negative impacts on performance.
#define LARGEST_REGEXP_OVERLAP    1024

#define SCALPEL_SIZEOFBUFFER_PANIC_STRING \
"PANIC: SIZE_OF_BUFFER has been incorrectly configured.\n"

#define SCALPEL_BLOCK_SIZE            512
#define MAX_STRING_LENGTH            4096
#define MAX_NEEDLES                   254
#define NUM_SEARCH_SPEC_ELEMENTS        6
#define MAX_SUFFIX_LENGTH               8
#define MAX_FILE_TYPES                100
#define MAX_MATCHES_PER_BUFFER        (SIZE_OF_BUFFER / 10)	// BUG: MUST ERROR OUT PROPERLY ON OVERFLOW (check)

// Length of the queues used to tranfer data / results blocks to workers.
#define QUEUELEN 20

#define MAX_FILES_PER_SUBDIRECTORY    1000

#define SCALPEL_OK                             0
#define SCALPEL_ERROR_NO_SEARCH_SPEC           1
#define SCALPEL_ERROR_FILE_OPEN                2
#define SCALPEL_ERROR_FILE_READ                3
#define SCALPEL_ERROR_FILE_WRITE               4
#define SCALPEL_ERROR_FILE_CLOSE               5
#define SCALPEL_ERROR_TOO_MANY_TYPES           6
#define SCALPEL_ERROR_FATAL_READ               7
#define SCALPEL_ERROR_BAD_HEADER_REGEX         8
#define SCALPEL_ERROR_BAD_FOOTER_REGEX         9
#define SCALPEL_ERROR_FILE_TOO_SMALL          10
#define SCALPEL_ERROR_NONEMPTY_DIRECTORY      11
#define SCALPEL_ERROR_PTHREAD_FAILURE         12

#define SCALPEL_GENERAL_ABORT                999

#define UNITS_BYTES                     0
#define UNITS_KILOB                     1
#define UNITS_MEGAB                     2
#define UNITS_GIGAB                     3
#define UNITS_TERAB                     4
#define UNITS_PETAB                     5
#define UNITS_EXAB                      6

// GLOBALS


// signal has been caught by signal handler
extern int signal_caught;

// current wildcard character
extern char wildcard;

// width of tty, for progress bar
extern int ttywidth;

extern int errno;

extern double totalsearch;	// # of seconds spent in pass # 1 header/footer searches
extern double totalqueues;	// # of seconds spent building work queues
extern double totalreads;	// # of seconds spent in all passes for input file reads
extern double totalwrites;	// # of seconds spent in pass # 2 for writing carved files

#define SCALPEL_NOEXTENSION_SUFFIX "NONE"
#define SCALPEL_NOEXTENSION '\xFF'

#define SCALPEL_DEFAULT_WILDCARD       '?'
#define SCALPEL_DEFAULT_CONFIG_FILE    "scalpel.conf"

#define SCALPEL_DEFAULT_OUTPUT_DIR     "scalpel-output"

#define SCALPEL_BANNER_STRING \
"Scalpel version %s\n"\
"Written by Golden G. Richard III and Lodovico Marziale.\n", SCALPEL_VERSION

#define SCALPEL_COPYRIGHT_STRING \
"Scalpel is (c) 2005-13 by Golden G. Richard III and Lodovico Marziale.\n"

// During the file carving operations (which occur after an initial
// scan of an image file to build the header/footer database), we want
// to read the image file only once more, sequentially, for all
// carves.  The following structure tracks the filename and first/last
// bytes in the image file for a single file to be carved.  When the
// read buffer includes the first byte of a file, the file is opened
// and the first write occurs.  When the read buffer includes the end
// byte, the last write operation occurs, the file is closed, and the
// struct can be reused.

// *****GGRIII: use of priority field to store these flags and the 
// data structures which track CarveInfo structs needs to be better
// documented

#define STARTCARVE      1	// carve operation for this CarveInfo struct
				// starts in current buffer
#define STOPCARVE       2	// carve operation stops in current buffer
#define STARTSTOPCARVE  3	// carve operation both starts and stops in
				// current buffer
#define CONTINUECARVE   4	// carve operation includes entire contents
				// of current buffer

typedef struct CarveInfo {
  char *filename;		// output filename for file to carve
  FILE *fp;			// file descriptor for file to carve
  unsigned long long start;	// offset of first byte in file
  unsigned long long stop;	// offset of last byte in file
  char chopped;			// is carved file's length constrained
  // by max file size for type? (i.e., could
  // the file actually be longer?
} CarveInfo;


// Each struct SearchSpecLine defines a particular file type,
// including header and footer information.  The following structure,
// SearchSpecOffsets, defines the absolute locations of all matching
// headers and footers for a particular file type.  Because the entire
// header/footer database is built during a single pass over an image
// or device file, the header and footer locations are sorted in
// ascending order.

typedef struct SearchSpecOffsets {
  unsigned long long *headers;	// positions of discovered headers
  size_t *headerlens;		// lengths of discovered headers
  unsigned long long headerstorage;	// space allocated for this many header offsets
  unsigned long long numheaders;	// # stored header positions
  unsigned long long *footers;	// positions of discovered footers
  size_t *footerlens;		// lengths of discovered footers
  unsigned long long footerstorage;	// space allocated for this many footer offsets
  unsigned long long numfooters;	// # stored footer positions
} SearchSpecOffsets;

// max files to open at once during carving--modify if you get
// a "too many files open" error message during the second carving phase.
#ifdef _WIN32
#define MAX_FILES_TO_OPEN            20
#else
#define MAX_FILES_TO_OPEN            512
#endif


typedef union SearchState {
  size_t bm_table[UCHAR_MAX + 1];
  regex_t re;
} SearchState;

typedef struct SearchSpecLine {
  char *suffix;
  int casesensitive;
  unsigned long long length;
  unsigned long long minlength;
  char *begin;          // translate()-d header
  char *begintext;      // textual version of header for humans
  int beginlength;
  int beginisRE;
  SearchState beginstate;
  char *end;            // translate()-d footer
  char *endtext;        // textual version of footer for humans
  int endlength;
  int endisRE;
  SearchState endstate;
  int searchtype;		// FORWARD, NEXT, REVERSE search type for footer
  struct SearchSpecOffsets offsets;
  unsigned long long numfilestocarve;	// # files to carve of this type
  unsigned long organizeDirNum;	// subdirectory # for organization 
  // of files of this type
} SearchSpecLine;


//prototype for external carving function
extern int scalpel_carveSingleInput(ScalpelInputReader * const reader,
		const char * const confFilePath,
		const char * const outDir,
		const unsigned char generateFooterDb,
		const unsigned char handleEmbedded,
		const unsigned char organizeSubdirs,
		const unsigned char previewMode,
		const unsigned char carveWithMissingFooters,
		const unsigned char noSearchOverlap
		) throw (std::runtime_error);

typedef struct scalpelState {
  ScalpelInputReader * inReader;
  char *conffile;
  char *outputdirectory;
  int specLines;
  struct SearchSpecLine *SearchSpec;
  unsigned long long fileswritten;
  int modeVerbose;
  int modeNoSuffix;
  FILE *auditFile;
  char *invocation;
  unsigned long long skip;
  char *coveragefile;
  unsigned int coverageblocksize;
  FILE *coverageblockmap;
  unsigned char *coveragebitmap;
  unsigned long long coveragenumblocks;
  int useInputFileList;
  char *inputFileList;
  int carveWithMissingFooters;
  int noSearchOverlap;
  int handleEmbedded;
  int generateHeaderFooterDatabase;
  int updateCoverageBlockmap;
  int useCoverageBlockmap;
  int organizeSubdirectories;
  unsigned long long organizeMaxFilesPerSub;
  int blockAlignedOnly;
  unsigned int alignedblocksize;
  int previewMode;
} scalpelState;


// one extent for a fragmented file.  'start' and 'stop'
// are real disk image addresses that define the fragment's
// location.
typedef struct Fragment {
  unsigned long long start;
  unsigned long long stop;
} Fragment;


// prototypes for visible scalpel.c functions
void freeState(struct scalpelState *state);
void initializeState(char ** argv, struct scalpelState *state);
void convertFileNames(struct scalpelState *state);
int readSearchSpecFile(struct scalpelState *state);


// prototypes for visible dig.c functions
int init_threading_model (struct scalpelState *state);
void destroy_threading_model(struct scalpelState *state);
int digImageFile (struct scalpelState *state);
int carveImageFile (struct scalpelState *state);
void init_store ();  // return int for error??
void destroyStore();

// prototypes for visible helpers.c functions

// LMIII fix me
#ifndef _WIN32
double elapsed (struct timeval A, struct timeval B);
//double elapsed(gettimeofday_t a, gettimeofday_t b);
#endif
int isRegularExpression (char *s);
void checkMemoryAllocation (struct scalpelState *state, void *ptr, int line,
			    const char *file, const char *structure);
int skipInFile (struct scalpelState *state, ScalpelInputReader * inReader);
void scalpelLog (struct scalpelState *state, const char *format, ...);
void handleError (struct scalpelState *s, int error);
int memwildcardcmp (const void *s1, const void *s2,
		    size_t n, int caseSensitive);
void setProgramName (char *s);
void init_bm_table (char *needle, size_t table[UCHAR_MAX + 1],
		    size_t len, int casesensitive);
int findLongestNeedle (struct SearchSpecLine *SearchSpec);
regmatch_t *re_needleinhaystack (regex_t * needle,
				 char *haystack, size_t haystack_len);
char *bm_needleinhaystack (char *needle, size_t needle_len,
			   char *haystack, size_t haystack_len,
			   size_t table[UCHAR_MAX + 1], int casesensitive);
int translate (char *str);
char *skipWhiteSpace (char *str);
void setttywidth ();

// prototypes for visible files.c functions
long long measureOpenFile (FILE * f, struct scalpelState *state);
int openAuditFile (struct scalpelState *state);
int closeAuditFile (FILE * f);

//// prototypes for visible dig.cu functions
int gpuSearchBuffer (char *readbuffer, int size_of_buffer, char *gpuresults,
		     int longestneedle, char wildcard);
void copytodevicepattern (char hostpatterntable[MAX_PATTERNS][MAX_PATTERN_LENGTH]);
void copytodevicelookup_headers(char hostlookuptable[LOOKUP_ROWS][LOOKUP_COLUMNS]);
void copytodevicelookup_footers(char hostlookuptable[LOOKUP_ROWS][LOOKUP_COLUMNS]);
void ourCudaMallocHost (void **ptr, int len);
int gpu_init (int longestneedle);
int gpu_cleanup();

// WIN32 string.h wierdness
#ifdef _WIN32
extern const char *strsignal (int sig);
//extern char *strsignal (int sig);
#else
extern char *strsignal (int sig);
#endif /*  ifdef _WIN32 */

#endif /* ifndef SCALPEL_H */
